File: Problem/Medium/401-500/424. Longest Repeating Character Replacement/__b.java
// Execution : 4 ms

// can use any loop ... make no changes

// use this
// char[] chars = s.toCharArray();
// then ++chars[right]
// then chars[left]++

// s.charAt(...) ... take more time

class Solution {
	public int characterReplacement(String s, int k) {
		int arr[] = new int[26];
		int n = s.length();

		int max_freq = 0, res = 0, window_len;

		int left = 0, right = 0;
		while (right < n) {
			max_freq = Math.max(max_freq, ++arr[s.charAt(right) - 'A']);
			window_len = right - left + 1;

			if (window_len - max_freq <= k) {
				res = Math.max(res, window_len);
			} else {
				arr[s.charAt(left) - 'A']--;
				left++;
			}
			right++;
		}

		return res;
	}
}

File: Problem/Medium/401-500/424. Longest Repeating Character Replacement/_b.py
# 85 ms
# res : max window length


class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count = {}
        n = len(s)

        window_len = max_freq = res = l = 0

        for r in range(n):
            window_len = r - l + 1
            ch = s[r]

            count[ch] = 1 + count.get(ch, 0)
            max_freq = max(max_freq, count[ch])

            if window_len - max_freq > k:
                count[s[l]] -= 1
                l += 1
            else:
                res = max(res, window_len)

                # 88 ms if use this
            # if window_len - max_freq <= k:
            #     res = max(res, window_len)
            # else:
            #     count[s[l]] -= 1
            #     l += 1

        return res
